### PR TITLE
Fix learning rate history and learning rate computation is schedulers

### PR DESCRIPTION
**Pull request by davidtvs**
See #42 for the discussion that led to this PR. Thanks @yongduek for bringing this issue to our attention.

@NaleRaphael would you kindly do a review of the changes here?

### CODE REVIEW

**Comment by davidtvs** [comment]
Bahhhh, github actions seems to only have PyTorch 1.4.0 and 1.5.0, so we can't run the tests on older versions 😞 .

Made a post on the community forum, hopefully I did something wrong or there's a workaround

**Comment by NaleRaphael** [comment]
Nice! Give me some time to do this!

**Comment by NaleRaphael** [comment]
I've run this revision on my machine with python-[3.6, 3.7] + torch-[0.4, 1.4], all works fine. 👍

And, good news! For the problem of installing older version of PyTorch, it's not that complicated as I imagined.

I used to install older version of PyTorch through the direct url found in [PyTorch archive][pytorch_archive]. But after re-visiting the official website of PyTorch to find some information, I found that we can just simply put this argument `-f https://download.pytorch.org/whl/torch_stable.html` in the `pip-install` command to solve this problem. 😅 (see also [this link][pytorch_install_older_version])


But there are a few problems:
1. There is no PyTorch-0.4.1 for Python-3.8, it's only available for Python-[2.7, 3.5, 3.6, 3.7]. You can check this by searching in the [PyTorch archive][pytorch_archive] with keyword `torch-0.4.1`. In order to run PyTorch on at least 2 different Python versions, I would suggest adding [`exclude`][github_actions_exclude] and [`include`][github_actions_include] keywords to manage it. (I created a repo for testing this, you can [check it out here][gh_actions_test_workflow])

2. With the [current setting][codecov_setting] for uploading code coverage report, there will be only one version uploaded to codecov. Hence that the report for code works on different versions of PyTorch might not be recorded. To solve this, you can just remove that condition to make each job upload the report, and codecov will handle it automatically.


Besides, would you like to make GitHub Actions be able to skip builds? In my opinion, it's a nice feature when we just want to update documentation without triggering CI. Travis-CI supports this feature by adding a line `skip ci` in commit message. And I found it is also possible in GitHub Actions and [had tested it][gh_actions_test_skip], see also [this coomment][gh_actions_skip_ci] for details.



[pytorch_archive]: https://download.pytorch.org/whl/torch_stable.html
[pytorch_install_older_version]: https://pytorch.org/get-started/previous-versions/#linux-and-windows-1
[codecov_setting]: https://github.com/davidtvs/pytorch-lr-finder/blob/7a2b5b8d6c85e9756495a8d60ad05e7cf10674b7/.github/workflows/ci_build.yml#L51
[github_actions_exclude]: https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-excluding-configurations-from-a-matrix
[github_actions_include]: https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-including-new-combinations
[gh_actions_test_workflow]: https://github.com/NaleRaphael/gh_actions_test/blob/3caf4f66c3305773cc4309a01b90391438f9cf4a/.github/workflows/ci_build.yml#L13-L21
[gh_actions_skip_ci]: https://github.community/t/github-actions-does-not-respect-skip-ci/17325/9
[gh_actions_test_skip]: https://github.com/NaleRaphael/gh_actions_test/actions/runs/119827519

File: .github/workflows/ci_build.yml [hunk 1]
```diff
@@ -21,24 +22,33 @@ jobs:
...
-          flake8 . --count --exit-zero --max-complexity=15 --max-line-length=127 --statistics
 
       - name: Run tests and generate coverage
         run: |
           pytest --cov=torch_lr_finder --cov-report=xml tests/
-          
+
       - name: Upload coverage to Codecov
-        if: matrix.python-version == 3.8
+        if: matrix.python-version == 3.8 && matrix.pytorch-version == '1.5.0'
```
Comment (NaleRaphael) [inline .github/workflows/ci_build.yml:51]: This line can be removed to make code coverage report of each job be uploaded to codecov.

File: .github/workflows/ci_build.yml [hunk 2]
```diff
@@ -11,7 +11,8 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.5, 3.6, 3.7, 3.8]
+        python-version: [3.5, 3.8]
+        pytorch-version: [0.4.1, 1.5.0]
```
Comment (NaleRaphael) [inline .github/workflows/ci_build.yml:15]: We can use `exclude` and `include` here to solve the problem mentioned in the comment.

```yaml
matrix:
  python-version: [3.5, 3.8]
  pytorch-version: [0.4.1, 1.5.0]
  exclude:
    - python-version: 3.8
      pytorch-version: 0.4.1
  include:
    - python-version: 3.7
      pytorch-version: 0.4.1
```

With this setting, jobs will be
```
ci-build (3.5, 0.4.1)
ci-build (3.5, 1.5.0)
ci-build (3.8, 1.5.0)
ci-build (3.7, 0.4.1)
```

File: .github/workflows/ci_build.yml [hunk 3]
```diff
@@ -21,24 +22,33 @@ jobs:
...
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-py-${{ matrix.python-version }}-pytorch-${{ matrix.pytorch-version }}-pip-${{ hashFiles('**/setup.py') }}
+          restore-keys: |
+            ${{ runner.os }}-py-${{ matrix.python-version }}-pytorch-${{ matrix.pytorch-version }}-pip-
+
+      - name: Install dependencies with PyTorch ${{ matrix.pytorch-version }}
+        if: steps.cache.outputs.cache-hit != 'true'
         run: |
           python -m pip install --upgrade pip
+          pip install torch==${{ matrix.pytorch-version }}
```
Comment (NaleRaphael) [inline .github/workflows/ci_build.yml:37]: Change this line to the following one can solve the problem of installing older version of PyTorch.

```bash
pip install torch==${{ matrix.pytorch-version }} -f https://download.pytorch.org/whl/torch_stable.html
```

**Comment by davidtvs** [comment]
Thanks for the help @NaleRaphael. I fixed the issues and added the "skip ci"

**Comment by NaleRaphael** [comment]
In [this run][gh_actions_723232868], there is a error "RuntimeError: Could not infer dtype of numpy.int64". But it's weird, it works fine on my machine...

lr_finder_test_run_py37_torch041

[gh_actions_723232868]: https://github.com/davidtvs/pytorch-lr-finder/runs/723232868?check_suite_focus=true

**Comment by davidtvs** [comment]
I could reproduce locally. But ya, it's weird, somehow all is fine with Pytorch 0.4.1 and Python 3.5 and 3.6; but with Python 3.7 I get the error ¯\_(ツ)_/¯

**Comment by NaleRaphael** [comment]
OK, though I find a similar issue [here](https://github.com/pytorch/pytorch/issues/8365), but it seems not worthy to dig into it.
I think we can merge this PR now. 👍

**Pull request merged**